# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [5]:
def addTwo(x):
    return x+2

print(addTwo(4))

#lambda

print((lambda x: x+2)(4))

a_num=4
    
print((lambda x: x+2)(a_num))

6
6
6


#### Saving to a Variable

In [6]:
f_test = lambda x: x+2

f_test(4)

6

#### Multiple Inputs

In [10]:
#multiple inputs with no variable

print((lambda x,y,z: x* y * z)(3,5,8))

120


#### Passing a Lambda into a Function

In [8]:
def multiply(f,num):
    return f(num)


multiply(lambda x: x*x, 4)

16

#### Returning a Lambda from a Function

In [14]:
def multiply_test(num):
    return num * 4

def returnfunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnfunc()
print(returnfunc())
print(f_return(8))


<function returnfunc.<locals>.multiply at 0x0000019B72893AF0>
16


#### If Statements within Lambdas

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [7]:

f = (lambda x: x**3)(4)

print(f)

64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [7]:
def squared(num,num2):
    if num < 10 and num2 < 10:
        return num **2, num2 ** 2
    else:
        return num, num2

numbers = [4,11,20,3,15,20]
more_nums = [4,10,3,2,6]

squared_nums_maps = list(map(squared,numbers,more_nums))
print(squared_nums_maps)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [10]:
squared_numbs_lambda = list(map(lambda x, y: (x **2, y**2) if x< 10 and y < 10 else (x,y), numbers, more_nums))
print(squared_numbs_lambda)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [14]:
squared_numbs_lambda = list(map(lambda x, y: (x * 2 -1, y*2-1), numbers, more_nums))
print(squared_numbs_lambda)

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [7]:
names = ['Bob', 'Andy', 'Max' , 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
    
new_names_list = list(filter(a_names, names))
print(new_names_list)
    

['Andy', 'Angelica']


#### Using Lambda's with Filter()

In [6]:

names = ['Bob', 'Andy', 'Max' , 'Evan', 'Angelica']


new_names_lambda = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))

print(new_names_lambda)


['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [ ]:
from statistics import mean
n = [2,3,4,5,6,7,8]

def a_names(name):
    if name <= mean(n):
        return True
    else:
        return False
    
new_names_list = list(filter(a_names, names))
print(new_names_list)

    


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [21]:
from functools import reduce
#reduce(function,iterable)


list_1=[2,4,6,8,10]
def addnums(num1,num2):
    return num1 + num2

result_add = reduce(addnums,list_1)
print(result_add)

def subtractnums(num1,num2):
    return num1-num2

result_sub = reduce(subtractnums,list_1)

print(result_sub)

30
-26


#### Using Lambda's with Reduce()

In [23]:
result_lamb = reduce(lambda x,y: x+y,list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [26]:
mylist = [1,2,3,4,5]
result_lamb2 = reduce(lambda x,y: x*y,mylist)
print(result_lamb2)

120


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [3]:
def addnums(num):
    if num <=1:
        print('addnums(1) = 1')
        return num
    else:
        print(f'addnums {num} - {num} + addnum({num -1})')
        return num + addnums(num -1)


addnums(10)

addnums 10 - 10 + addnum(9)
addnums 9 - 9 + addnum(8)
addnums 8 - 8 + addnum(7)
addnums 7 - 7 + addnum(6)
addnums 6 - 6 + addnum(5)
addnums 5 - 5 + addnum(4)
addnums 4 - 4 + addnum(3)
addnums 3 - 3 + addnum(2)
addnums 2 - 2 + addnum(1)
addnums(1) = 1


55

#### Writing a Factorial Function

In [6]:
def factorial(num):
    if num <=1:
        return num
    else:
        return num * factorial(num-1)
    
print(factorial(12))

479001600


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [3]:
def subtract(num):
    if num <=1:
        return num
    else:
        return num - subtract(num-1)
    
print(subtract(1))

1


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [18]:
def my_range(stop,start,step = 1):
    while start > stop:
        yield start 
        start -= step
for i in my_range(10,start = 10):
    my_generator_value = i 
    print(my_generator_value)

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [32]:
def my_range(start):
    while start > 0:
        yield start 
        start -= 1
for i in my_range(start = 10):
    my_generator_value = i ** 2
    print(my_generator_value)
    
    

    

100
81
64
49
36
25
16
9
4
1
<generator object my_range at 0x0000025BA1432820>


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [2]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

list2 = ['', ' ' ,'  ','    ','     ']
output_list = [i for i in places if i not in list2]
 
print(str(output_list))

    



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [3]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

print(sorted(author, key=lambda x: x.title().split(" ")[-1]))



['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [6]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
newplaces = list(map(lambda c:  (c[0], (9/5) * c[1]+ 32), places))
print((newplaces))


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [7]:
def Iteration(xn):
    if xn == 0:
        return 0
    elif xn == 1:
        return 1
    else:
        return Iteration(xn - 1) + Iteration(xn - 2)

def fib(n):
    for i in range(n + 1):
        print("Iteration({i}): {r}".format(i=i,r=Iteration(i+1)))

fib(5)


Iteration(0): 1
Iteration(1): 1
Iteration(2): 2
Iteration(3): 3
Iteration(4): 5
Iteration(5): 8
